File: php/functions/QueryShipment.php
<?php
/**
 * Yurtiçi Kargo - QueryShipment (Gönderi Sorgulama)
 * 
 * Müşterinin kendi referans kodu ile gönderi geçmişini raporlayan servis.
 * 
 * @requires PHP 8.0+, ext-soap
 */

class QueryShipment
{
    private SoapClient $soapClient;
    private string $username;
    private string $password;
    private string $language;

    /** Key tipleri */
    public const KEY_TYPE_CARGO = 0;
    public const KEY_TYPE_INVOICE = 1;

    /** Operasyon durum kodları */
    public const STATUS_NOP = 0;
    public const STATUS_IND = 1;
    public const STATUS_ISR = 2;
    public const STATUS_CNL = 3;
    public const STATUS_ISC = 4;
    public const STATUS_DLV = 5;
    public const STATUS_BI  = 6;

    public function __construct(SoapClient $soapClient, string $username, string $password, string $language = 'TR')
    {
        $this->soapClient = $soapClient;
        $this->username = $username;
        $this->password = $password;
        $this->language = $language;
    }

    /**
     * Gönderi sorgulama yapar.
     *
     * @param array $keys Sorgulanacak anahtar değerleri
     * @param int $keyType 0=cargoKey, 1=invoiceKey
     * @param bool $addHistoricalData Hareket geçmişi dahil edilsin mi
     * @param bool $onlyTracking Sadece takip URL'si dönsün mü
     * @return object Response object
     * @throws SoapFault
     * @throws InvalidArgumentException
     */
    public function execute(array $keys, int $keyType = 0, bool $addHistoricalData = false, bool $onlyTracking = false): object
    {
        if (empty($keys)) {
            throw new \InvalidArgumentException('En az bir anahtar değeri gereklidir.');
        }

        if (!in_array($keyType, [0, 1], true)) {
            throw new \InvalidArgumentException('keyType 0 (cargoKey) veya 1 (invoiceKey) olmalıdır.');
        }

        $keysStr = implode(',', array_map('strval', $keys));

        $params = new \stdClass();
        $params->wsUserName = $this->username;
        $params->wsPassword = $this->password;
        $params->wsLanguage = $this->language;
        $params->keys = $keysStr;
        $params->keyType = $keyType;
        $params->addHistoricalData = $addHistoricalData;
        $params->onlyTracking = $onlyTracking;

        $response = $this->soapClient->__soapCall('queryShipment', [$params]);

        return $this->parseResponse($response);
    }

    /**
     * SOAP yanıtını parse eder.
     */
    private function parseResponse(object $response): object
    {
        $result = new \stdClass();

        $data = $response->ShippingDeliveryVO ?? $response;

        $result->outFlag = (int) ($data->outFlag ?? 2);
        $result->outResult = (string) ($data->outResult ?? '');
        $result->senderCustId = (int) ($data->senderCustId ?? 0);
        $result->count = (int) ($data->count ?? 0);
        $result->details = [];

        if (isset($data->shippingDeliveryDetailVO)) {
            $details = is_array($data->shippingDeliveryDetailVO)
                ? $data->shippingDeliveryDetailVO
                : [$data->shippingDeliveryDetailVO];

            foreach ($details as $detail) {
                $item = new \stdClass();
                $item->cargoKey = (string) ($detail->cargoKey ?? '');
                $item->invoiceKey = (string) ($detail->invoiceKey ?? '');
                $item->jobId = (int) ($detail->jobId ?? 0);
                $item->docId = (string) ($detail->docId ?? '');
                $item->operationCode = (int) ($detail->operationCode ?? 0);
                $item->operationMessage = (string) ($detail->operationMessage ?? '');
                $item->operationStatus = (string) ($detail->operationStatus ?? '');
                $item->errCode = (int) ($detail->errCode ?? 0);
                $item->errMessage = (string) ($detail->errMessage ?? '');
                $item->itemDetail = null;

                if (isset($detail->shippingDeliveryItemDetailVO)) {
                    $itemDetail = $detail->shippingDeliveryItemDetailVO;
                    $item->itemDetail = new \stdClass();
                    $item->itemDetail->trackingUrl = (string) ($itemDetail->trackingUrl ?? '');
                    $item->itemDetail->receiverCustName = (string) ($itemDetail->receiverCustName ?? '');
                    $item->itemDetail->departureUnitName = (string) ($itemDetail->departureUnitName ?? '');
                    $item->itemDetail->deliveryDate = (string) ($itemDetail->deliveryDate ?? '');
                    $item->itemDetail->deliveryTime = (string) ($itemDetail->deliveryTime ?? '');
                    $item->itemDetail->totalDesi = (float) ($itemDetail->totalDesi ?? 0);
                    $item->itemDetail->cargoHistory = [];

                    if (isset($itemDetail->invDocCargoVOArray)) {
                        $events = is_array($itemDetail->invDocCargoVOArray)
                            ? $itemDetail->invDocCargoVOArray
                            : [$itemDetail->invDocCargoVOArray];

                        foreach ($events as $event) {
                            $item->itemDetail->cargoHistory[] = [
                                'unitName'  => (string) ($event->unitName ?? ''),
                                'eventName' => (string) ($event->eventName ?? ''),
                                'reasonName' => (string) ($event->reasonName ?? ''),
                                'eventDate' => (string) ($event->eventDate ?? ''),
                                'eventTime' => (string) ($event->eventTime ?? ''),
                                'cityName'  => (string) ($event->cityName ?? ''),
                                'townName'  => (string) ($event->townName ?? ''),
                            ];
                        }
                    }
                }

                $item->isSuccess = fn() => $item->errCode === 0;
                $result->details[] = $item;
            }
        }

        $result->isSuccess = fn() => $result->outFlag === 0;

        return $result;
    }

    /**
     * Durum kodunun açıklamasını döndürür.
     */
    public static function getStatusDescription(int $code): string
    {
        return match ($code) {
            self::STATUS_NOP => 'İşlem yapılmadı, düzenlenmemiş',
            self::STATUS_IND => 'Kargo teslimattadır',
            self::STATUS_ISR => 'Düzenlenmiş, fatura prosedürü tamamlanmamış',
            self::STATUS_CNL => 'Çıkış engellendi',
            self::STATUS_ISC => 'İptal edilmiş',
            self::STATUS_DLV => 'Teslim edildi',
            self::STATUS_BI  => 'YK acente tarafından iptal edildi',
            default => 'Bilinmeyen durum kodu',
        };
    }
}

File: java/src/main/java/com/yurticikargo/functions/QueryShipment.java
package com.yurticikargo.functions;

import com.yurticikargo.models.QueryResult;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.io.ByteArrayInputStream;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;

/**
 * queryShipment SOAP fonksiyonu.
 * Müşterinin kendi referans kodu ile gönderi geçmişini raporlayan servistir.
 */
public class QueryShipment {

    private static final String SOAP_NS = "http://yurticikargo.com.tr/ShippingOrderDispatcherServices";

    private final String endpoint;
    private final String username;
    private final String password;
    private final String language;
    private final HttpClient httpClient;

    public QueryShipment(String endpoint, String username, String password, String language, HttpClient httpClient) {
        this.endpoint = endpoint;
        this.username = username;
        this.password = password;
        this.language = language;
        this.httpClient = httpClient;
    }

    /**
     * Gönderi sorgulama isteği gönderir.
     *
     * @param keys              Sorgulanacak anahtarlar
     * @param keyType           0=CargoKey, 1=InvoiceKey
     * @param addHistoricalData Hareket geçmişi dahil mi?
     * @param onlyTracking      Sadece takip URL'si mi?
     * @return QueryResult
     */
    public QueryResult execute(String[] keys, int keyType, boolean addHistoricalData, boolean onlyTracking) throws Exception {
        String soapXml = buildRequest(keys, keyType, addHistoricalData, onlyTracking);
        String responseXml = sendRequest(soapXml);
        return parseResponse(responseXml);
    }

    private String buildRequest(String[] keys, int keyType, boolean addHistoricalData, boolean onlyTracking) {
        StringBuilder sb = new StringBuilder();
        sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
        sb.append("<soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\" ");
        sb.append("xmlns:ship=\"").append(SOAP_NS).append("\">\n");
        sb.append("  <soapenv:Header/>\n");
        sb.append("  <soapenv:Body>\n");
        sb.append("    <ship:queryShipment>\n");
        sb.append("      <wsUserName>").append(escapeXml(username)).append("</wsUserName>\n");
        sb.append("      <wsPassword>").append(escapeXml(password)).append("</wsPassword>\n");
        sb.append("      <wsLanguage>").append(escapeXml(language)).append("</wsLanguage>\n");
        // WSDL: maxOccurs='unbounded' - each key as separate element
        for (String key : keys) {
            sb.append("      <keys>").append(escapeXml(key)).append("</keys>\n");
        }
        sb.append("      <keyType>").append(keyType).append("</keyType>\n");
        sb.append("      <addHistoricalData>").append(addHistoricalData).append("</addHistoricalData>\n");
        sb.append("      <onlyTracking>").append(onlyTracking).append("</onlyTracking>\n");
        sb.append("    </ship:queryShipment>\n");
        sb.append("  </soapenv:Body>\n");
        sb.append("</soapenv:Envelope>");
        return sb.toString();
    }

    private String sendRequest(String soapXml) throws Exception {
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(endpoint))
                .header("Content-Type", "text/xml; charset=utf-8")
                .header("SOAPAction", "")
                .POST(HttpRequest.BodyPublishers.ofString(soapXml, StandardCharsets.UTF_8))
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
        return response.body();
    }

    public QueryResult parseResponse(String xml) throws Exception {
        QueryResult result = new QueryResult();

        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        factory.setNamespaceAware(false);
        DocumentBuilder builder = factory.newDocumentBuilder();
        Document doc = builder.parse(new ByteArrayInputStream(xml.getBytes(StandardCharsets.UTF_8)));

        // Check for SOAP Fault
        NodeList faultNodes = doc.getElementsByTagName("faultstring");
        if (faultNodes.getLength() > 0) {
            result.setOutFlag(2);
            result.setOutResult(faultNodes.item(0).getTextContent().trim());
            return result;
        }

        result.setOutFlag(getIntValue(doc, "outFlag", 2));
        result.setOutResult(getTextValue(doc, "outResult"));

        // shippingDeliveryDetailVO
        NodeList detailNodes = doc.getElementsByTagName("shippingDeliveryDetailVO");
        for (int i = 0; i < detailNodes.getLength(); i++) {
            Element detailEl = (Element) detailNodes.item(i);
            QueryResult.QueryDetail detail = new QueryResult.QueryDetail();
            detail.setCargoKey(getTextValue(detailEl, "cargoKey"));
            detail.setInvoiceKey(getTextValue(detailEl, "invoiceKey"));
            detail.setOperationCode(getIntValue(detailEl, "operationCode", 0));
            detail.setOperationMessage(getTextValue(detailEl, "operationMessage"));
            detail.setOperationStatus(getTextValue(detailEl, "operationStatus"));
            detail.setErrCode(getIntValue(detailEl, "errCode", 0));
            detail.setErrMessage(getTextValue(detailEl, "errMessage"));

            // shippingDeliveryItemDetailVO
            NodeList itemNodes = detailEl.getElementsByTagName("shippingDeliveryItemDetailVO");
            if (itemNodes.getLength() > 0) {
                Element itemEl = (Element) itemNodes.item(0);
                QueryResult.QueryItemDetail itemDetail = new QueryResult.QueryItemDetail();
                itemDetail.setTrackingUrl(getTextValue(itemEl, "trackingUrl"));
                itemDetail.setReceiverCustName(getTextValue(itemEl, "receiverCustName"));
                itemDetail.setReceiverAddress(getTextValue(itemEl, "receiverAddressTxt"));
                itemDetail.setDepartureUnitName(getTextValue(itemEl, "departureUnitName"));
                itemDetail.setDeliveryDate(getTextValue(itemEl, "deliveryDate"));
                itemDetail.setDeliveryTime(getTextValue(itemEl, "deliveryTime"));

                // invDocCargoVOArray (cargo history)
                NodeList historyNodes = itemEl.getElementsByTagName("invDocCargoVOArray");
                for (int j = 0; j < historyNodes.getLength(); j++) {
                    Element histEl = (Element) historyNodes.item(j);
                    QueryResult.CargoEvent event = new QueryResult.CargoEvent();
                    event.setUnitName(getTextValue(histEl, "unitName"));
                    event.setEventName(getTextValue(histEl, "eventName"));
                    event.setEventDate(getTextValue(histEl, "eventDate"));
                    event.setEventTime(getTextValue(histEl, "eventTime"));
                    event.setReasonName(getTextValue(histEl, "reasonName"));
                    itemDetail.getCargoHistory().add(event);
                }

                detail.setItemDetail(itemDetail);
            }

            result.getDetails().add(detail);
        }

        return result;
    }

    private String getTextValue(Element parent, String tagName) {
        NodeList list = parent.getElementsByTagName(tagName);
        if (list.getLength() > 0 && list.item(0).getTextContent() != null) {
            return list.item(0).getTextContent().trim();
        }
        return "";
    }

    private String getTextValue(Document doc, String tagName) {
        NodeList list = doc.getElementsByTagName(tagName);
        if (list.getLength() > 0 && list.item(0).getTextContent() != null) {
            return list.item(0).getTextContent().trim();
        }
        return "";
    }

    private int getIntValue(Element parent, String tagName, int defaultVal) {
        String val = getTextValue(parent, tagName);
        if (!val.isEmpty()) {
            try { return Integer.parseInt(val); } catch (NumberFormatException e) { /* ignore */ }
        }
        return defaultVal;
    }

    private int getIntValue(Document doc, String tagName, int defaultVal) {
        String val = getTextValue(doc, tagName);
        if (!val.isEmpty()) {
            try { return Integer.parseInt(val); } catch (NumberFormatException e) { /* ignore */ }
        }
        return defaultVal;
    }

    private String escapeXml(String value) {
        if (value == null) return "";
        return value.replace("&", "&amp;")
                    .replace("<", "&lt;")
                    .replace(">", "&gt;")
                    .replace("\"", "&quot;")
                    .replace("'", "&apos;");
    }

    /**
     * Debug: SOAP XML isteğini döndürür.
     */
    public String getRequestXml(String[] keys, int keyType, boolean addHistoricalData, boolean onlyTracking) {
        return buildRequest(keys, keyType, addHistoricalData, onlyTracking);
    }
}
